### PR TITLE
DOC: major documentation rewrite

### LINKED ISSUE #138: Document support for installing python scripts
How can I install python scripts with `meson-python` replacing setuptool's `console_scripts` `enty_point` or `[project.scripts]` in `pyproject.toml`? If there is a vision of how this should look like I'm happy to try to work on the support.

### LINKED ISSUE #197: Rework documentation
The documentation is not super accessible to new users. I would like to rewrite it so that it is more accessible to new users, while still presenting the required information for advanced users.

### LINKED ISSUE #230: config-settings support needs docs
This is a follow-up issue for gh-167. The most important thing to add is docs. The invocations are verbose and argument quoting can be difficult, so we need a good number of one-liners as usage examples (there are some near the bottom of gh-167).

Also, it's possible that there's still an issue with functionality. `--config-settings` doesn't take multiple arguments at once, so if you'd want to pass multiple arguments to `meson setup`, it seems like there's no good way to do it. Something like:
```
python -m pip install . --no-build-isolation --config-settings='setup-args=-Dblas=blas' --config-settings='setup-args=-Dlapack=lapack'
```
seems to just use one of the two arguments. @tacaswell and I tried for a while, and didn't get it to work. It's possible we missed something.

### LINKED ISSUE #253: Add the PyPA discord as an official project communication channel
I have created a `meson-python` channel in the "Packaging Related" (meaning it's affiliated with the PyPA) category of the PyPA discord. Is everyone okay in documenting this as one of the official communication channels for the project?

https://discord.gg/pypa

### LINKED ISSUE #264: Consider dropping direct references to PEP 517
See https://discuss.python.org/t/name-for-pyproject-toml-builds-and-peps/1659 for the rationale.

Basically, we've been moving to a world where "pyproject.toml-based builds" is a much more consistent and better way to refer to such build-backends. We don't need to point users to dense design documents to understand how stuff works.

### PR DESCRIPTION
**Pull request by FFY00**
Closes #197, closes #230, closes #253, and closes #264
Sets us up to fix later #138, #233, and #224

Signed-off-by: Filipe Laíns <lains@riseup.net>

### CODE REVIEW

**Comment by FFY00** [comment]
Sorry for the big PR, I'd recommend you having a look to see if there is anything major wrong. If there isn't anything major wrong, we should probably merge, and then we can send followup PRs to improve it.

File: docs/tutorials/introduction.rst [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+   If you are not familiar with the TOML configuration language or need some
+   help, be sure to check out
+   `their documentation <https://toml.io/en/v1.0.0>`__.
+
+
+Generating the distribution artifacts
+=====================================
+
+Now that we have a valid Python project, we can release it.
```
Comment (rgommers) [inline docs/tutorials/introduction.rst:175]: There is one issue here I think (also related to the FIXME 3 paragraphs down): you go straight from metadata to a release, without local use first. How about inserting a section _"Building and installing our package"_ first? Touching on `pip install .`, mentioning that builds a wheel for the current Python interpreter and then installs it. Show that it's installed with `pip list`, then import it. Conclude that we now have a working package. Next step: a release.

File: docs/tutorials/introduction.rst [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+.. admonition:: Check the pypa/build documentation
+   :class: seealso
+
+   To learn more about `pypa/build`_, you can check
+   `their documentation <https://pypa-build.readthedocs.io/>`__.
+
+
+.. admonition:: Building wheels for many systems
+   :class: tip
+
```
Comment (rgommers) [inline docs/tutorials/introduction.rst:240]: I'd say here that for a pure Python package, all one needs is one sdist and one wheel for a release. So we'd be good at this point to upload to PyPI. For a package with an extension module, one needs multiple wheels - they are specific to the OS, hardware, and (unless one uses the stable ABI) the Python minor version. So `.. warning::` this must be done with `cibuildwheel`.

File: docs/how-to-guides/add-dependencies.rst [hunk 3]
```diff
@@ -0,0 +1,13 @@
+.. _how-to-guides-add-dependencies:
+
+*******************
+Adding dependencies
```
Comment (rgommers) [inline docs/how-to-guides/add-dependencies.rst:4]: Is this page needed? At least it should start by stating that this is all standardized, and nothing is specific to `meson-python`. It looks like the Python Packaging User Guide has no good section on this - maybe we should fix that (after finishing up this PR) so we can link to it for more details.

File: docs/how-to-guides/build-directory.rst [hunk 4]
```diff
@@ -0,0 +1,14 @@
+.. _how-to-guides-build-directory:
+
+*******************************
+Use a permanent build directory
+*******************************
+
+
+.. todo::
+
+   - Explain why you want to use the ``builddir`` option
```
Comment (rgommers) [inline docs/how-to-guides/build-directory.rst:10]: This includes things that are hard to pass through a front end's config-settings I think. Would be good to point out that this works as expected when you, for example, run `meson setup --cross-file ...`, and `meson-python` will reconfigure only what's needed.

File: docs/how-to-guides/config-settings.rst [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+On `pip`_, they can be used via the ``--config-settings`` option.
+
+
+.. code-block:: console
+
+   $ pip install . --config-settings key=value
+
+
+Please check the `pypa/build documentation`_ for more information.
```
Comment (rgommers) [inline docs/how-to-guides/config-settings.rst:39]: typo: `pip` docs, not `build` docs

File: docs/how-to-guides/editable-installs.rst [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+This will install the project in the current directory in editable mode. When a
+``meson-python`` package is installed in editable mode, it will be re-built the
+first time it is imported. That means, the first time you import a module from
+your package, it will take a bit longer.
+
+
+.. admonition:: What is the scope of editable installs?
+   :class: seealso
+
+   The editable installs feature only targets the Python modules of the project.
```
Comment (rgommers) [inline docs/how-to-guides/editable-installs.rst:25]: This sounds like pure Python. It's worth pointing out that, unlike in any other build backend, files for an extension module (C/C++/Cython/whatever) are also editable and will be automatically rebuilt.

File: docs/how-to-guides/editable-installs.rst [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+you simply need to set the ``MESONPY_EDITABLE_VERBOSE`` environment variable
+to any non-null value.
+
+
+.. code-block:: console
+
+   $ MESONPY_EDITABLE_VERBOSE=1 python
+
+
+Enable it permanently
```
Comment (rgommers) [inline docs/how-to-guides/editable-installs.rst:54]: Just deleting these section headers will make the page look better. If it's one sentence + one line of code, that's a sign that it's too little for its own section.

File: docs/how-to-guides/editable-installs.rst [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+
+.. admonition:: ``MESONPY_EDITABLE_VERBOSE`` will have no effect here
+   :class: attention
+
+   Please note that the ``MESONPY_EDITABLE_VERBOSE`` environment variable will
+   not have **any** effect on the installation, it simply forces verbose mode to
+   be enabled **after** the package has been installed.
+
+
```
Comment (rgommers) [inline docs/how-to-guides/editable-installs.rst:76]: The limitations when working with an IDE (breakpoints, running tests from the UI) are worth pointing out here. And saying work is planned to improve this.

File: docs/how-to-guides/executable-with-internal-dependencies.rst [hunk 9]
```diff
@@ -0,0 +1,20 @@
+.. _how-to-guides-executable-with-internal-dependencies:
+
+*************************************
+Executable with internal dependencies
+*************************************
+
+
+.. todo::
+
```
Comment (rgommers) [inline docs/how-to-guides/executable-with-internal-dependencies.rst:9]: The macOS PR for this is about ready to merge, and we can probably sort Windows out soon. So maybe the static library explanation isn't needed?

File: docs/how-to-guides/editable-installs.rst (hunk 7 above)
Comment (rgommers) [reply docs/how-to-guides/editable-installs.rst:54]: This applies to other pages as well. For example, the _Use build config settings_ does not need section headers for `pypa/build` and `pip`.

File: docs/how-to-guides/meson-args.rst [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+***************************
+Passing arguments to Meson_
+***************************
+
+
+.. todo::
+
+   - Explain how you can add extra Meson arguments via ``pyproject.toml``
+   - Explain how you can add extra Meson arguments via the build conifig
```
Comment (rgommers) [inline docs/how-to-guides/meson-args.rst:11]: minor: typo in last word here

File: docs/reference/limitations.rst [hunk 11]
```diff
@@ -0,0 +1,50 @@
+.. _reference-limitations:
+
+***********
+Limitations
+***********
+
```
Comment (rgommers) [inline docs/reference/limitations.rst:6]: The issue with the `license` key in `pyproject.toml` is probably worth pointing out.

**Review by rgommers — COMMENTED** [review]
Thanks @FFY00! This looks like a good structure. I added some comments from a first read.

Content suggestions I have:
- a how-to for controlling what files get included in an sdist (with `.gitattributes`) and that a change to a vcs-tracked file gets included but a new file does not. It's mentioned briefly in the section on sdist, but could use more detail and an example `.gitattributes file
- a tutorial page on cross-compiling (I can fill it in, I am dealing with bits and pieces of this in multiple places already)
- a how-to page on including a VCS-generated version (e.g. with `fs.exists` snippet in https://github.com/mesonbuild/meson-python/issues/159#issuecomment-1265962676)
- A _Who uses meson-python_ page, linked from the front page.

**Comment by rgommers** [comment]
> Sorry for the big PR, I'd recommend you having a look to see if there is anything major wrong. If there isn't anything major wrong, we should probably merge, and then we can send followup PRs to improve it.

It seems like a good strategy to merge when the structure is right (it matches the suggestion in gh-197; let's give everyone a couple of days to have at least a high-level look at this) and what is there in terms of text is in decent shape. So we get the improved docs for the issues that this already closes. And then we can parallelize filling in the most important blanks.

**Comment by rgommers** [comment]
By the way, on updating this PR: it would be quite useful to not rewrite history as you go. That makes it impossible to check what changed for reviewers. Rather, push new commits as updates, and then squash commits right at the end.

**Comment by FFY00** [comment]
> By the way, on updating this PR: it would be quite useful to not rewrite history as you go. That makes it impossible to check what changed for reviewers. Rather, push new commits as updates, and then squash commits right at the end.

Yeah, sorry. Github isn't a great git review platform. It would be awesome if they added support for rebase-merges with autosquash, that way I could push fixup commits for specific commit, and not have to squash the whole history. Currently, to update a specific commit I need to force-push, with the only alternative being committing to squash the whole history on merge. FYI you can actually click in "force-pushed" in the message to see what changed.

I want to push some of your feedback as separate commits, so instead of force-pushing, I can just push the fixup commits without squashing for review, and then when it comes time to merge, I will squash them and merge. Would that work for you? The other, more annoying, alternative would be to just push those changes as a separate PR.

File: docs/tutorials/introduction.rst [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+we want it to build, and how to do it.
+
+
+.. code-block:: meson
+   :caption: meson.build
+
+   project('purelib-and-platlib', 'c')
+
+   py_mod = import('python')
+   py = py_mod.find_installation()
```
[the remark this answers is not in the document]
Comment (FFY00) [reply docs/tutorials/introduction.rst:87]: From https://mesonbuild.com/Python-module.html#find_installation

>  `pure`: On some platforms, architecture independent files are expected to be placed in a separate directory. However, if the python sources should be installed alongside an extension module built with this module, this keyword argument can be used to override the default behavior of `.install_sources()`. since 0.64.0 

Hummm... I do not follow. AFAICT this will just force sources to be placed in platlib.

File: docs/tutorials/introduction.rst [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+
+.. admonition:: Writing TOML
+   :class: seealso
+
+   If you are not familiar with the TOML configuration language or need some
+   help, be sure to check out
+   `their documentation <https://toml.io/en/v1.0.0>`__.
+
+
+Generating the distribution artifacts
```
[the remark this answers is not in the document]
Comment (FFY00) [reply docs/tutorials/introduction.rst:172]: > Maybe this is too technical a title. How about _"Creating a first release of your package"_?

I changed it to "Creating your first release" :+1:

File: docs/how-to-guides/add-dependencies.rst (hunk 3 above)
Comment (FFY00) [reply docs/how-to-guides/add-dependencies.rst:4]: I didn't want to get into it in the tutorial, but wanted to have somewhere to point to quickly learn this. I'd say keep it for now, try to fix the Packaging User Guide, and then update this based on the outcome.

File: docs/how-to-guides/build-directory.rst (hunk 4 above)
Comment (FFY00) [reply docs/how-to-guides/build-directory.rst:10]: Could you elaborate? I don't think I follow :sweat_smile:

File: docs/how-to-guides/editable-installs.rst (hunk 7 above)
Comment (FFY00) [reply docs/how-to-guides/editable-installs.rst:54]: Yeah :sweat_smile:. I wanted to use smaller headers, but rST doesn't allow that.

File: docs/how-to-guides/executable-with-internal-dependencies.rst (hunk 9 above)
Comment (FFY00) [reply docs/how-to-guides/executable-with-internal-dependencies.rst:9]: Well, it's already written :rofl:. We can just keep it and remove when it's fixed, I don't think it'd be much trouble.

File: docs/how-to-guides/meson-args.rst (hunk 10 above)
Comment (FFY00) [reply docs/how-to-guides/meson-args.rst:11]: Yeah, I also caught that locally when I added spellchecking to the docs, which I will push in a separate commit. I was originally thinking of doing a different PR, but it'll have conflicts and it's pretty simple anyway, so I will probably push it here instead.

**Comment by FFY00** [comment]
I fixed the smaller stuff you mentioned, I will do the rest monday :+1:

File: docs/who-uses-meson-python.rst [hunk 14]
```diff
@@ -0,0 +1,26 @@
+.. _who-uses-meson-python:
+
+**************************
+Who uses ``meson-python``?
+**************************
```
Comment (dnicolodi) [inline docs/who-uses-meson-python.rst:5]: I don't like adding a list of projects using meson-python to the documentation, or anywhere else. As we don't have any need for advertisement (and I'm pretty sure we don't want to be advertising any other project) it does not serve any purpose and it is going to be a burden to maintain. For the users, it is much better to have a short and curated list of projects that follow best practices and that can be used as examples.

Unless the listed projects are known to following best practices, the list is not going to be useful for the users. I don't think we have resources to review all the projects submitted to be listed, and to keep our review current.

I would much prefer it this would be something like:
```
 Curated list of projects using ``meson-python``
===================================

The following projects use Meson as their build system and meson-python to produce Python wheels. They are roughly listed in order of complexity:

- `siphash24 <https://github.com/dnicolodi/siphash24>`_ A very simple project. It demonstrates how Meson makes it trivial to compile a CPython extension written in Cython via a simple tempate engine and link it to a library compiled from a Meson subproject. Also an example of how to use `cibuildwheel`_ to produce Python wheels for several platforms.

- (i don't know the peculiarities of the other projects)

- `SciPy <https://github.com/scipy/scipy>`_ Probably the most complex Python project using Meson and ``meson-python``. Staring at the build definition too long may cause headaches. 

```

File: docs/tutorials/introduction.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply docs/tutorials/introduction.rst:87]: Which is what you want, since you have C code in your package. This is a safety measure; otherwise you have to plaster `pure: false` everywhere you use `py.install_sources`, and if you forget it just once you end up with a broken package. That's why it was added: https://mesonbuild.com/Release-notes-for-0-64-0.html#pythonfind_installation-now-accepts-pure-argument

[the remark this answers is not in the document]
Comment (dnicolodi) [reply docs/tutorials/introduction.rst:87]: Why do you end up with a broken package? Is there an assumption somewhere that all the code in a package needs to be installed either in `{platlib}` or in `{purelib}`? meson-python does not do this assumption, and actually has code to handle this case. If this is the case, it would be nice to detect the circumstance and raise an error in meson-python. However, I don't see why this is the case. Also, in most platforms, `{platlib}` and `{purelib}` are the same path.

File: docs/how-to-guides/build-directory.rst (hunk 4 above)
Comment (rgommers) [reply docs/how-to-guides/build-directory.rst]: I've come across this multiple times already, for cross-compilation in particular. See for example https://github.com/scipy/scipy/blob/1c836efe5ff37ffa4490756269b060a464690e62/.github/workflows/wheels.yml#L180-L188.

Also for conda-forge, you need in `build.sh` something like this for cross-compiling (the first line is how the conda-forge build tooling configures automatically for cross-compilation):
```
meson ${MESON_ARGS} builddir
python -m build -Cbuilddir=builddir
```

File: docs/tutorials/introduction.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply docs/tutorials/introduction.rst:87]: > Is there an assumption somewhere that all the code in a package needs to be installed either in `{platlib}` or in `{purelib}`?

Yes of course. Say you have a `foo.py` file and `bar.so` extension module, and you do a relative import `from .bar` in `foo.py`, that only works if the two files are in the same directory. Which is the kind of thing you only see is broken when some niche Linux distro doesn't have purelib and platlib pointing to the same path (not Ubuntu, which is typically the one thing people test in their CI). This whole purelib/platlib concept is just awful.

> meson-python does not do this assumption, and actually has code to handle this case. If this is the case, it would be nice to detect the circumstance and raise an error in meson-python.

Not every build goes only through `meson-python`. If it's broken for plain `meson setup/install`, that's still bad. Raising an error SGTM.

[the remark this answers is not in the document]
Comment (dnicolodi) [reply docs/tutorials/introduction.rst:87]: Ah, that's obvious, but I can have a wheel that installs multiple top-level modules. And those can be installed some in `{platlib}` and some in `{purelib}`. Saying that if you don't pass `pure: false` you get a broken wheel is a simplification that I don't think we should make.

[the remark this answers is not in the document]
Comment (rgommers) [reply docs/tutorials/introduction.rst:87]: Multiple top-level modules is a corner case that we're not dealing with here. I'm not asking to say _" if you don't pass `pure: false` you get a broken wheel "_. I am asking to please use the robust `pure: false` as the default everywhere, instead of the very fragile state of not doing that. Which I'm quite sure is the right thing to do (for more details, please see https://github.com/mesonbuild/meson/issues/10523).

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply docs/tutorials/introduction.rst:87]: Setting this on the installation object will flip the default value on the install_sources() and get_install_dir() functions. This is the correct, sensible, sane, and *only* reasonable thing to do if you have a top-level module with, nested inside it, a mixture of pure python and compiled extensions: it means that by default, you install pure python together with the compiled extension.

It doesn't prevent you from passing `pure: true` to individual calls where you do intend to install some freestanding top-level pure-python modules to purelib anyways. And I agree, that's a valid use case -- although I expect it's also *incredibly* rare.

Failing to place pure python files in the platlib when they are nested inside a package together with platlib extensions, means that you end up with one package being a namespace package and the other not, with corresponding... abuses... of the importlib search path. You are not going to end up happy if that happens, because at least one user is going to be on a distro which splits the two, and then disaster is going to happen and it's not going to be pretty.

We should absolutely document the default recommended approach, as the one that doesn't break people. :)

**Comment by FFY00** [comment]
The force-push was from a rebase to solve conflicts. The commits being reviewed haven't been changed other than just solving conflicts.

As mentioned previously, updates to the original commits are being done via fixup commits. They will be squashed after approval, just before merging.

File: docs/who-uses-meson-python.rst (hunk 14 above)
Comment (FFY00) [reply docs/who-uses-meson-python.rst:9]: Are the new updates good for you, or do you change the wording further?

File: docs/tutorials/introduction.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply docs/tutorials/introduction.rst:91]: I think it's reasonable to set that, but IMO it would be best to not set it, and add an admonition below pointing to a page that explains this in-depth.

> Failing to place pure python files in the platlib when they are nested inside a package together with platlib extensions, means that you end up with one package being a namespace package and the other not, with corresponding... abuses... of the importlib search path. You are not going to end up happy if that happens, because at least one user is going to be on a distro which splits the two, and then disaster is going to happen and it's not going to be pretty.

We should try to detect these cases and warn users.

[the remark this answers is not in the document]
Comment (FFY00) [reply docs/tutorials/introduction.rst:91]: I added this

```rst
.. admonition:: Mixing pure and native code
   :class: caution

   If you are mixing pure (``.py``) and native modules in your project, please
   check out :ref:`reference-quirks-mixing-purelib-and-platlib`, which explains
   the common foot-guns when doing so, and how to make sure you install your
   modules correctly.
```

**Comment by FFY00** [comment]
@rgommers I think I've address all your comments, could you do a second pass?

File: docs/tutorials/introduction.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply docs/tutorials/introduction.rst:100]: @FFY00, I disagree - this is not a choice between two equivalent alternatives. `pure: false` is so far better, it must be used. You have no idea how much time this has already cost me in subtle bugs, and how tricky it is to debug. Please just accept the comment. Keeping a bad default and then adding cognitive load for the user with a note explaining why they should not follow your example code because of footguns is not useful.

File: docs/explanations/design.rst [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+*******************************
+
+
+.. todo::
+
+   - Explain how ``meson-python`` maps files to the wheel
+
+     - Link to :ref:`tutorials-executable` in the relevant section
+
+   - Explain why the heuristics are necessary and how they work
```
Comment (rgommers) [inline docs/explanations/design.rst:18]: Aren't we about to remove those heuristics? Or are these not the ones for `{moduledir_shared}` & co?

File: docs/how-to-guides/build-directory.rst [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+Use a permanent build directory
+*******************************
+
+
+.. todo::
+
+   - Explain why you'd want to use the ``builddir`` option
+     - Mention it works properly when using ``--cross-file`` (eg. https://github.com/scipy/scipy/blob/1c836efe5ff37ffa4490756269b060a464690e62/.github/workflows/wheels.yml#L180-L188)
+   - Explain how ``builddir`` works
+   - Warn that user-configured build directories are not "use at your own risk"
```
Comment (rgommers) [inline docs/how-to-guides/build-directory.rst:17]: I think the intent here was "are" rather tha" are not" (at your own risk)? That said, I'd rephrase as "Explain the constraints that the user must adhere to when using a user-configured build directory". That also takes care of the second warning the line below.

File: docs/reference/limitations.rst [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+interpreter instances at the same time is not supported.
+
+
+Using editable installs with IDEs
+=================================
+
+Currently, with editable installs, setting breakpoints via an IDE or similar
+tool will not work.
+
+We have work planed to fix this issue.
```
Comment (rgommers) [inline docs/reference/limitations.rst:38]: minor: typo in "planed"

File: docs/tutorials/introduction.rst [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+
+As you might have figured out by the name, *sdists* contain the source code of
+the project, and *wheels* contain a compiled [#pure-wheels]_ version of the
+project, ready to be copied to the file-system.
+
+If your project has compiled code, its *wheels* will be platform-dependent, and
+if it has native Python modules, the *wheels* will also be Python version
+specific, as the native Python modules depend on the Python ABI.
+FIXME(this seems *a bit* convoluted, it isn't awful, but I feel like it could
+be simplified or explained in a more natural way)
```
Comment (rgommers) [inline docs/tutorials/introduction.rst:238]: This still needs addressing. I'd start with the common case:
```
If your project uses Python extension modules, your *wheels* will be specific to
both the platform and the Python version (footnote: unless you use CPython's stable ABI).
```

I would not even bother with a case where there is native code, but no extension modules - this is quite unlikely to happen (it'd require using `ctypes` or `cffi` to do anything with the native code).

File: docs/tutorials/introduction.rst [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+
+Strictly speaking, you don't need to distribute *wheels*, however they make the
+user experience much nicer. Unless you have any reason not to, we highly
+recommend you distribute *wheels* for at least the most common systems. When
+*wheels* are not available for a system, the project can still be installed, be
+it needs to be build from the *sdist*, which involves fetching all the build
+dependencies and going through the likely expensive build.
+FIXME(strictly speaking -- this doesn't sound great, how do we highlight the
```
Comment (rgommers) [inline docs/tutorials/introduction.rst:246]: How about: 
```
While distributing *wheels* is not mandatory, they make the
user experience much nicer. Unless you have any reason not to, we highly
```

File: docs/who-uses-meson-python.rst [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+Here's a curated list of projects using ``meson-python``.
+
+
+.. list-table::
+   :widths: 20 80
+
+   * - `SciPy <https://github.com/scipy/scipy>`_
+     - Probably the most complex project using Meson and ``meson-python``.
+       Staring at the build definition too long may cause headaches.
```
Comment (rgommers) [inline docs/who-uses-meson-python.rst:19]: I'm not sure if I should confirm that this is true, or be slightly offended:)

File: docs/who-uses-meson-python.rst [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+
+   * - `scikit-image <https://github.com/scikit-image/scikit-image>`_
+     - Another complex project using ``meson-python``.
```
Comment (rgommers) [inline docs/who-uses-meson-python.rst:22]: I suggest: "A project of medium complexity, with a lot of Cython code"

(no need to repeat "using meson-python" - that's above the table already)

**Review by rgommers — COMMENTED** [review]
This is pretty close. I suggested some rewordings for the FIXMEs.

File: docs/who-uses-meson-python.rst (hunk 20 above)
Comment (eli-schwartz) [reply docs/who-uses-meson-python.rst:19]: Anyone who gets headaches from looking at SciPy's build definition should definitely not look at the systemd build definition, is all I can say. :)

File: docs/explanations/design.rst (hunk 15 above)
Comment (FFY00) [reply docs/explanations/design.rst:18]: We are, this was written before that PR.

File: docs/how-to-guides/build-directory.rst (hunk 16 above)
Comment (FFY00) [reply docs/how-to-guides/build-directory.rst:17]: Yes. Thank you for catching that.

File: docs/tutorials/introduction.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply docs/tutorials/introduction.rst]: I changed it.

IMO this is the kind of issue we can, and should be trying to detect.

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @FFY00

File: docs/who-uses-meson-python.rst (hunk 14 above)
Comment (FFY00) [reply docs/who-uses-meson-python.rst:9]: @dnicolodi please let me know if you'd like further changes, I think this is the only blocker for merging this.

File: docs/how-to-guides/editable-installs.rst [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+Verbose mode
+============
+
+It might be useful for you to see the output of Meson_ when the package is being
+re-built. We provide a *verbose mode* that does this, and can be enabled either
+temporarily, after the install, or permanently, in the install.
+
+
+To enable the verbose mode on any existing ``meson-python`` editable install,
+you simply need to set the ``MESONPY_EDITABLE_VERBOSE`` environment variable
```
Comment (dnicolodi) [inline docs/how-to-guides/editable-installs.rst:51]: It is not related to this PR but it came to my attention scrolling through the added documentation: I think that the `mesonpy` name should not be used and that we should stick with `meson-python` everywhere. Thus I would prefer if we standardize on `MESON_PYTHON_` as the prefix for environment variable names.

File: docs/who-uses-meson-python.rst (hunk 20 above)
Comment (dnicolodi) [reply docs/who-uses-meson-python.rst:19]: I didn't expect it to be taken seriously when I wrote that sentence. Unless everyone is happy with the pun, I would replace with something a bit more useful. Something like: "It combines CPython extensions and libraries written in C, C++, Fortran, Cython, Pythran, different ASM variants, ..." maybe something about CPU feature detection or BLAS libraries. I'm sure @rgommers can suggest something.

File: docs/who-uses-meson-python.rst [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+**************************
+Who uses ``meson-python``?
+**************************
+
+Here's a curated list of projects using ``meson-python``.
+
+
+.. list-table::
+   :widths: 20 80
```
Comment (dnicolodi) [inline docs/who-uses-meson-python.rst:15]: I haven't see how this is rendered, but I don't know if a table is the best choice. Maybe a simple itemized list is better.

File: docs/who-uses-meson-python.rst (hunk 14 above)
Comment (dnicolodi) [reply docs/who-uses-meson-python.rst:9]: I'm not completely happy with the section name. I would prefer something like "Projects using meson-python".

File: docs/who-uses-meson-python.rst [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+   * - `scikit-image <https://github.com/scikit-image/scikit-image>`_
+     - A project of medium complexity, with a lot of Cython_ code.
+
+   * - `siphash24 <https://github.com/dnicolodi/python-siphash24>`_
+     - A very simple project. It demonstrates how Meson makes it trivial to
+       compile a `CPython extension`_ written in `Cython`_ via a simple template
+       engine and link it to a library compiled from a Meson sub-project. Also
+       an example of how to use `cibuildwheel`_ to produce Python wheels for
+       several platforms.
```
Comment (dnicolodi) [inline docs/who-uses-meson-python.rst:29]: I used this little project of mine just because I know it, I'm not sure it is a good example. Feel free to pick another one.

**Review by dnicolodi — APPROVED** [review]
I had a look at that section only. I have some suggestions on the wording, but nothing major. I have a comment unrelated to this PR that we can discuss separately.

File: docs/who-uses-meson-python.rst (hunk 23 above)
Comment (FFY00) [reply docs/who-uses-meson-python.rst:15]: I tried that, but preferred the list.

File: docs/who-uses-meson-python.rst (hunk 24 above)
Comment (FFY00) [reply docs/who-uses-meson-python.rst:29]: There aren't *that* many projects out there right now. I originally added the main ones that seemed alright to me. Anyway, it is good to have a simpler one as an example.

**Comment by FFY00** [comment]
I pushed a final update, [here](https://github.com/mesonbuild/meson-python/pull/276/commits/8ab7fbdfcae8b0d01b72a5191a9d6eef359c32f6). I am not rebase and merge now.

@rgommers please do send a PR re-writing the SciPy description if you feel anything can be improved there.

**Comment by rgommers** [comment]
> @rgommers please do send a PR re-writing the SciPy description if you feel anything can be improved there.

Okay, will do.
